### PR TITLE
#5260: remote query support for AA bots

### LINKED ISSUE #5260: Request throttled when fetch bot list from AA Control Room
Context
---
- https://pixiebrix.slack.com/archives/C042T2TBZ17/p1677258501309459

Discussion
---
- This might occur if the user has a lot of bots available to them because we loop over all the files available in order to populate the list
- To avoid the throttling, might need to pass a search query

Related Code
---
- https://github.com/pixiebrix/pixiebrix-extension/blob/de11b1b7886413fc9ec72c19ad918fbebc87aafb/src/contrib/automationanywhere/BotOptions.tsx#L162-L162
- https://github.com/pixiebrix/pixiebrix-extension/blob/de11b1b7886413fc9ec72c19ad918fbebc87aafb/src/contrib/automationanywhere/aaApi.ts#L167-L167

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #5260 
- Adds a AsyncRemoteSelectWidget component that queries results on demand
- Switches the AA Bot field to use AsyncRemoteSelectWidget

## Remaining Work

- [x] Add affordance for looking up the default option when it's not defined in the initial set of options: the affordance is that the factory method is provided with the current value, so it can decide how it wants to search based on the current query and value

## Demo

- [x] https://www.loom.com/share/f5d86f63628946b9b9c9b67aa7391910

## Future Work

- Decide where to switch other fields over to using the async widget

## Checklist

- [x] Add tests
- [x] Designate a primary reviewer: @BLoe

### CODE REVIEW

File: src/components/form/widgets/SelectWidget.tsx [hunk 1]
```diff
@@ -95,7 +95,7 @@ const SelectWidget = <TOption extends Option<TOption["value"]>>({
   };
 
   // Pass null instead of undefined if options is not defined
-  const selectValue =
+  const selectedOption =
```
Comment (twschiller) [inline src/components/form/widgets/SelectWidget.tsx:98]: Improved grammar

**Review by BLoe — APPROVED** [review]
I'm approving to unblock this, but have 2 real questions about promise syntax and why we need to be putting hacks in the test code for this ticket (time sensitivity?)

File: src/components/form/widgets/AsyncRemoteSelectWidget.tsx [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+          callback([
+            {
+              value: "error",
+              label: getErrorMessage(error, "Error loading options"),
+              isDisabled: true,
+            } as any,
+          ]);
+        });
+    },
+    75,
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/components/form/widgets/AsyncRemoteSelectWidget.tsx:105]: > 75ms seems like a very short debounce timeout

The 75ms in the debounce refers to how long does it wait after the user finishes typing before the query is sent. The maxWaitMillis refers to the maximum amount of time to wait before sending the request

If you don't debounce it, when I've previously used react-select, it runs on every keypress/change of the query

File: src/components/form/widgets/AsyncRemoteSelectWidget.test.tsx [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    await waitForEffect();
+    expect(optionsFactoryMock).not.toHaveBeenCalled();
+
+    await act(async () => {
+      await userEvent.click(container.querySelector('[role="combobox"]'));
+
+      await userEvent.type(container.querySelector('[role="combobox"]'), "foo");
+
+      // Wait for the debounce. :shrug: would be cleaner to advance timers here, but this will do for now
+      await sleep(100);
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/components/form/widgets/AsyncRemoteSelectWidget.test.tsx:140]: > Is there an issue/blocker with using fake timers here? Just sucks to explicitly add runtime to our tests just to wait for things...

There's no blockers, I want just pressed for time and have not used the fake timers before. I didn't get a chance to dig into the example here: https://github.com/pixiebrix/pixiebrix-extension/wiki/Unit-tests#dealing-with-async-state

**Comment by twschiller** [comment]
@BLoe I'm going to merge in, and then circle back on testing and seeing if the async/await call works later this week

File: src/components/form/widgets/AsyncRemoteSelectWidget.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (BLoe) [reply src/components/form/widgets/AsyncRemoteSelectWidget.tsx:105]: Gotcha, makes sense 👍

File: src/components/form/widgets/AsyncRemoteSelectWidget.tsx [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  value,
+  ...selectProps
+}) => {
+  const [knownOptions, setKnownOptions] = useState<Option[]>([]);
+
+  // `react-select` doesn't automatically debounce requests
+  const loadOptions = useDebouncedCallback(
+    (query: string, callback: (options: Option[]) => void) => {
+      void optionsFactory(config, { ...factoryArgs, query, value })
+        // eslint-disable-next-line promise/prefer-await-to-then -- https://stackoverflow.com/a/64773351
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/components/form/widgets/AsyncRemoteSelectWidget.tsx:85]: I didn't try using async/await. Will give that a try when I circle back on the testing to use fake timers

**Pull request merged**